### ISSUE TITLE
NVDA Should report alignment in LibreOffice Writer

### ISSUE THREAD

**Issue by nvaccessAuto** [nvaccess/nvda#2204]
**Reported by ateu on 2012-03-31 17:24**
Hello,

This is the only one information which is not reported in the IBM's Office Suite.
Please furgive me if htis is a problem with IBM Lothus Symphony.

Regardes
**Blocking #4003**

**Comment by ehollig** [comment]
@jcsteh, is this issue specific to Symphony or does it apply to OO/LO as well? If it is specific to Symphony then I suspect that this issue can be closed.

**Comment by jcsteh** [comment]
Applies to LO as well.

**Comment by Marlin86** [comment]
Unfortunately this problem persists. I think LibreOffice and OpenOffice are so important programs, as they are used by millions of people, public schools, Universities and other. Therefore, they should receive more attention.

**Comment by Adriani90** [comment]
I guess if Libre Office or Open Office do not expose this information, NVDA cannot do much to improve it. However, since they are also open source projects, I think people facing such issues could create issues on Github or on their issue trackers. I mean especially people who are using NVDA. I wonder if there are some issues reported by NVDA users on Libre Office or Open Office issue trackers?
Here is a wiki with all channels where you can report issues in Libre office:
https://wiki.documentfoundation.org/QA/BugReport

And here is how to submit issues in Open Office:
https://www.openoffice.org/qa/issue_handling/pre_submission.html

Regarding Open office, it seems the development is not very active anymore. There is like a new version in one or two years.

**Comment by surfer0627** [comment]
@Adriani90 commented 
> I guess if Libre Office or Open Office do not expose this information, NVDA cannot do much to improve it.

Yes. 
In the past, developer Jteh used to point out:
> Symphony actually doesn't follow the IAccessible2 standard for formatting info, .....

And [tdf 91739](https://bugs.documentfoundation.org/show_bug.cgi?id=91739) – nvda can not read documents from top to bottom. - is One of issues reported on LibreOffice issue tracker.

My question is that is it possible for NVDA developers to submit changes to the LibreOffice Git repository to fix IAccessible2 standard issue?

Related links:
* [Start developing LibreOffice! Registering with Git and Gerrit, and local settings](https://blog.documentfoundation.org/blog/2019/06/03/start-developing-libreoffice-registering-with-git-and-gerrit-and-local-settings/)

**Comment by joanmarie** [comment]
I came across this while trying to figure out some history about when text alignment started being treated as an object attribute in IA2. (It's a text attribute in ATK so we have API misalignment which makes implementors sad. But I digress.)

In LibreOffice Writer, for ATK, text alignment is correctly exposed as a text attribute. See https://cgit.freedesktop.org/libreoffice/core/tree/vcl/unx/gtk3/a11y/gtk3atktextattributes.cxx. In particular, this line:
```     "ParaAdjust",           // TEXT_ATTRIBUTE_JUSTIFICATION```

To see if that was happening in Windows as well, I installed LibreOffice, centered a paragraph, and used NVDA's Python Console to get that paragraph's text attributes. Here's what I got:

```
(0, 11, 'Version:1;:;CharBackColor:#00FFFFFF;CharColor:#00000000;CharContoured:0;CharEmphasis:0;CharEscapement:0;CharFontName:Liberation Serif;CharHeight:12;CharPosture:0;CharShadowed:0;CharStrikeout:0;CharUnderline:0;CharUnderlineColor:#00000000;CharWeight:100;ParaAdjust:3;ParaBottomMargin:0;ParaFirstLineIndent:0;ParaLeftMargin:0;ParaLineSpacing:Mode=0,Height=100,;ParaRightMargin:0;ParaTabStops:Position=48,TabAlign=0,DecimalChar= ,FillChar= ,;')
```

Bit of a blob, but note that `ParaAdjust:3` is there. That seems to be your alignment. It's not exposed as an object attribute, but it is obtainable as a text attribute. 

If NVDA is getting all the other text attributes in order to report them (e.g. in response to NVDA+f), it could arguably check for alignment as part of it. Just like I'm probably going to have to modify Orca to check the object attributes on top of the text attributes, because Firefox stopped treating text alignment as a text attribute. <insert shrug here>

**Comment by jcsteh** [comment]
> I came across this while trying to figure out some history about when text alignment started being treated as an object attribute in IA2.

As far as I know, it was always an object attribute in IA2. The justification was probably that it applies to the whole paragraph, not an individual span of text.

> Bit of a blob, but note that `ParaAdjust:3` is there.

Indeed, as I noted in https://github.com/nvaccess/nvda/issues/2204#issuecomment-155296609.

> If NVDA is getting all the other text attributes in order to report them (e.g. in response to NVDA+f), it could arguably check for alignment as part of it.

Oh, sure it could. NVDA already has a bunch of code to deal with LO's non-standard attribute names. Someone just has to implement it.

> Just like I'm probably going to have to modify Orca to check the object attributes on top of the text attributes, because Firefox stopped treating text alignment as a text attribute.

Off-topic, but you should file a Mozilla bug about that. if the documented ATK standard is to expose it as a text attribute, servers should follow the standard. :)

**Comment by michaelweghorn** [comment]
> Comment 7 by jteh on 2014-03-21 00:03 I discovered Symphony/OOo/LO expose this using the ParaAdjust text attribute, which is non-standard. (This isn't surprising, since all of their attributes are non-standard.) It seems that center is ParaAdjust:3. I'm not sure what the other values are yet.

LibreOffice 24.2 will use standard IA2 text and object attributes and with NVDA 2024.1, text alignment is reported as expected, s. https://bugs.documentfoundation.org/show_bug.cgi?id=135922 and PR https://github.com/nvaccess/nvda/pull/15649 for details.

I think this issue can therefore be closed. (And from what I can see, it's about the same issue as #6913.)

**Comment by Adriani90** [comment]
Thanks for updating and for the great work on this @michaelweghorn. Very appreciated.

I am closing this issue as works for me. Please comment if you are still having this issue in LO 24.2 and we can reopen.
